src/NuGet.Core/NuGet.Build.Tasks/StaticGraphRestoreTaskBase.cs

```diff
@@ -173,35 +220,69 @@ public override bool Execute()
         /// <summary>
         /// Gets the command-line arguments to use when launching the process that executes the restore.
         /// </summary>
-        internal IEnumerable<string> GetCommandLineArguments()
+        /// <param name="globalProperties">Receives a <see cref="Dictionary{TKey, TValue}" /> containing the global properties.</param>
+        internal string GetCommandLineArguments(Dictionary<string, string> globalProperties)
         {
+            // First get the command-line arguments including the global properties
+            string commandLineArguments = CreateArgumentString(EnumerateCommandLineArguments(SerializeGlobalProperties ? null : globalProperties));
+
+            if (SerializeGlobalProperties)
+            {
+                return commandLineArguments;
+            }
+
+            // If the command-line arguments exceed the supported length, set the flag to serialize the global properties to the standard input stream get the
+            // arguments again without the global properties.
+            if (commandLineArguments.Length > 8000)
```

Comment:
Can you walk me through the steps to get to 8,000?

[This](https://learn.microsoft.com/en-us/troubleshoot/windows-client/shell-experience/command-line-string-limitation#more-information) says the limit is 8,191 characters.

Is this 8,000 characters for just the global properties?

What about the full path to the executable and any other arguments?

I kinda want to see a formal calculation in the comment that accounts for everything and is clear about what is in and out of scope for the limit.  As it stands, I have a hard time seeing and validating the assumptions behind this magic number.

Reply:
I had read some conflicting information on the limits.  The documentation you pointed to is for a user typing a command-line into `cmd.exe`.  Other information says that the limit for `CreateProcess` itself is `USHORT / 2` or `32767`.  But since this is a console application, I think its going to be 8,191 on Windows.  I was having a tougher time figuring out the limit for other platforms.

So I settled on 8,000 as a safe limit.  I also wasn't sure if the limit even applies to the executable itself?

At the moment if the actual limit is exceeded, the process just crashes.  At least with these changes the user will get an error message.  If you feel confident in a particular limit, I'd be happy to use that.